[AN-Issue-1629] Optimize on_new_epoch to reduce gas-usage

aregng opened the pull request:
- created optimized version of on_new_epoch and try_withdraw_task_automation_fee
   - Now refund and task clean-up is done in one go if refund in applicable.
   - During epoch fee withdrawal of the tasks all the funds are collected and then deposited to registry account address in total
- Includes fix for [Automation] Potential Risk of Abusive Usage of Automation Registry Feature
[#1823](https://github.com/Entropy-Foundation/smr-moonshot/issues/1823)
    - Implemented deposit amount charge for the tasks during registration
    - Implemented deposit refund logic when tasks are cancelled, expired or stopped.
- Includes fix for [Automation] Events on task activation and removal for registry
[#1788](https://github.com/Entropy-Foundation/smr-moonshot/issues/1788)
   - Implemented events to be emitted related to active and removed tasks on epoch change
- Updated existing unit tests and added new ones to check new functionality


Latest results of the run
```
- No refund simply activating all tasks, 1 epoch
│             Test Name                                                    │ Time(secs) │ Gas Used│
│ 0x1::automation_registry::check_task_activation_on_new_epoch_performance │   0.290    │ 17267    │
│ 0x1::automation_registry::check_task_registration_performance            │   0.152    │ 9675     │


- Refund on epoch start, 2nd epoch

│ 0x1::automation_registry::check_task_activation_on_new_epoch_performance │   0.492    │ 31896    │
│ 0x1::automation_registry::check_task_registration_performance            │   0.147    │ 9675     │

- All tasks are expired, 3rd epoch
│ 0x1::automation_registry::check_task_activation_on_new_epoch_performance │   0.667    │ 47094    │
│ 0x1::automation_registry::check_task_registration_performance            │   0.142    │ 9675     │

```

With local run I have the following results:
```
- No refund simply activating all tasks, 1 epoch
│             Test Name                                                             │ Time(secs) │ Gas Used│
│ 0x1::automation_registry::check_task_activation_on_new_epoch_2_performance        │   1.216    │  80907  │                                                                                               
│ 0x1::automation_registry::check_task_activation_on_new_epoch_performance          │   0.858    │  82193  │
│ 0x1::automation_registry::check_task_registration_performance                     │   0.165    │  9382   │


- Refund on epoch start, 2nd epoch

│ 0x1::automation_registry::check_task_activation_on_new_epoch_2_performance        │   1.703    │ 156289  │
│ 0x1::automation_registry::check_task_activation_on_new_epoch_performance          │   1.817    │ 159968  │
│ 0x1::automation_registry::check_task_registration_performance                     │   0.195    │  9382   │

- All tasks are expired, 3rd epoch
│ 0x1::automation_registry::check_task_activation_on_new_epoch_2_performance        │   1.528    │ 162800  │
│ 0x1::automation_registry::check_task_activation_on_new_epoch_performance          │   1.214    │ 165396  │
│ 0x1::automation_registry::check_task_registration_performance                     │   0.143    │  9382   │

```

Commit:
[AN-Issue-1629] Optimize on_new_epoch to reduce gas-usage

- created optimized version of on_new_epoch and try_withdraw_task_automation_fee(s) for initial review

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -678,6 +707,151 @@ module supra_framework::automation_registry {
 
                 // Drop or activate task for this current epoch.
                 if (task.expiry_time <= current_time || task.state == CANCELLED) {
+                    coin::transfer<SupraCoin>(&resource_signer, task.owner, task.locked_fee_for_next_epoch);
```
@aregng, I believe the governance is not allowed to withdraw funds from the resource account so that there will never be any deficit in the resource account for refund, correct? Otherwise this `transfer` may fail during epoch transition and cause stalling the chain.

aregng replied:
Yes Dr. @sjoshisupra , when deposit is withdrawn from user account during registration, that amount will be added to the locked resource and withdrawal of the resource account from governance will be guarded to not allow over-withdrawal.

sjoshisupra replied:
@aregng , now that we have added `locked_fee_for_next_epoch` we should ensure that this amount is also guarded and should be prevented from withdrawal by governance.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -678,6 +707,151 @@ module supra_framework::automation_registry {
+        // only after refund has been done for previous epoch.
+        update_config_from_buffer();
+
+        // If feature is not enabled then we are not charging and tasks are cleared.
```
Are we assuming that if the `feature` is disabled, we would refund to all the task owners? Where/when is that refund happening? @aregng

aregng replied:
Yes, at line 760 I have TODO: to refund deposits before clean-up.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -678,6 +707,151 @@ module supra_framework::automation_registry {
+
+        vector::for_each(ids, |task_index| {
+            if (!enumerable_map::contains(&automation_registry.tasks, task_index)) {
+                event::emit(ErrorTaskDoesNotExist { task_index })
```
Given that `ids` is a result of `get_map_list`, is the containment check required here?

aregng replied:
yes, agree this check can be removed, assuming that `enumerable_map` is bug-free and additions and removals are handled and tested.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -867,50 +1047,169 @@ module supra_framework::automation_registry {
+        };
+        if (task.fee != 0) {
+            // Charge the fee and emit a success event
+            coin::transfer<SupraCoin>(
```
I think we discussed that it makes sense to use `coin::withdraw` and store it in a `Coin<SupraCoin>` and accumulate all such fee in a `Coin` and then use a single `coin::deposit` in the resource account to reduce storage operations.

Similarly, since refunds will have to be processed, perhaps it makes sense to withdraw entire belance in a `Coin` and then use `coin::extract` from that  and `coin::deposit` in user account to process refund and then at the end use a single `coin::deposit` back to the resource account.

aregng replied:
I remember this agreement. My initial investigation showed that `Coin<CoinType>` can be created only via `coin::withdraw` method, and for the `refund` context I think it is not suitable to use. Even if we collect first entire amount from `registry-account` during first iteration, then we should do second iteration to extract and deposit to each account, so I thought 2 iterations over single iteration and multiple withdraws will not give much performance.
But I will check and let you know,
As for the automation fee collection on new epoch, I will apply it. somehow I missed fee collection part and was concentrated on refunds more.

sjoshisupra replied:
@aregng , ok for refund perhaps using `extract` and then `deposit` may not save us much gas, I agree. but how about charging?

aregng replied:
will do.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -1181,6 +1480,22 @@ module supra_framework::automation_registry {
     }
 
+    /// Sorting vector implementation
+    public fun sort(input: &mut vector<u64>) {
+        let len = vector::length(input);
+        let i = 0;
+        while (i < len) {
+            let j = i + 1;
+            while (j < len) {
+                if (*vector::borrow(input, i) > *vector::borrow(input, j)) {
+                    vector::swap(input, i, j)
+                };
+                j = j + 1;
+            };
+            i = i + 1;
+        };
```
It might be more efficient to use insertion sort algorithm instead of selection sort. Insertion sort will perform better when the list is almost sorted, which is likely to be the case when we retrieve IDs from the map. @aregng

Commit:
Added full support of automation fee deposit and refund for tasks and updated sort to be insertion

nizam-supraoracles commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -266,6 +317,16 @@ module supra_framework::automation_registry {
 
+    #[event]
+    /// Event emitted when on new epoch refunds to be paied is not possible due to insufficient resource account balance.
+    /// Type of the refund can be either depoit payed during registration (0), or caused by the shortenting of the epoch (1)
```
`paied` -> `paid`
`depoit` -> `deposit`
`shortenting` -> `shortening`
I thing this one also `payed`

nizam-supraoracles commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -281,15 +342,40 @@ module supra_framework::automation_registry {
         fee: u64,
     }
 
+    /// Represents the fee charged for an automation task execution and some additional information.
+    struct AutomationTaskFeeMeta has drop {
+        task_index: u64,
+        owner: address,
+        fee: u64,
+        automation_fee_cap: u64,
+        expiry_time: u64,
+        max_gas_amount: u64,
+        locked_fee_for_next_epoch: u64
+    }
+
     /// Represents intermediate state of the registry on epoch change.
     struct IntermediateState has drop {
-        active_task_ids: vector<u64>,
+        removed_task_ids: vector<u64>,
```
Is the change is backward compatible ?

sjoshisupra replied:
If this `IntermediateState` struct existed earlier, this is not backward compatible. Please refer to [this article](https://aptos.dev/en/build/smart-contracts/book/package-upgrades) to see what care must be taken during upgrade/redeployment. We can not change any existing struct.

aregng replied:
I thought as long as it is internal struct without store capability, renaming the field is possible.
I will double check though.

aregng replied:
It turns out that even internal structures can not be modified, except adding new capabilities to the struct.

nizam-supraoracles commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -678,14 +813,255 @@ module supra_framework::automation_registry {
+        let result = safe_refund(resource_signer, resouce_address, task_index, task_owner, refundable_deposit, DEPOSIT_EPOCH_FEE);
+        event::emit(
+            TaskDepositFeeRefund { task_index, owner: task_owner, amount: refundable_deposit }
+        );
```
Isn't the event supposed to be emitted only if the result is true?

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -281,15 +342,40 @@ module supra_framework::automation_registry {
+            epoch_locked_fees,
+
+        } = value;
+        coin::deposit(automation_registry.registry_fee_address, coin::extract_all(&mut epoch_locked_fees));
+        coin::destroy_zero(epoch_locked_fees)
```
```suggestion
        coin::deposit(automation_registry.registry_fee_address, epoch_locked_fees);
        
```

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -328,12 +414,25 @@ module supra_framework::automation_registry {
 
+    #[view]
+    /// Get locked balance of the resource account in terms of deposited automation fees.
+    public fun get_locked_deposite_balance(): u64 acquires AutomationRefundBookkeeping {
```
```suggestion
    public fun get_locked_deposit_balance(): u64 acquires AutomationRefundBookkeeping {
```

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -328,12 +414,25 @@ module supra_framework::automation_registry {
+    #[view]
+    /// Get total locked balance of the resource account.
+    public fun get_registry_total_locked_balance(): u64 acquires AutomationRefundBookkeeping, AutomationRegistry {
+        get_epoch_locked_balance() + get_locked_deposite_balance()
```
```suggestion
        get_epoch_locked_balance() + get_locked_deposit_balance()
```

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -678,14 +813,255 @@ module supra_framework::automation_registry {
+        (tcmg, removed_tasks)
+    }
+
+    fun safe_refund(resource_signer: &signer, resouce_address: address, task_index: u64, task_owner: address, refundable_amount: u64, refund_type: u8):  bool {
```
```suggestion
    fun safe_refund(resource_signer: &signer, resource_address: address, task_index: u64, task_owner: address, refundable_amount: u64, refund_type: u8):  bool {
```

sjoshisupra replied:
`refund_type` is unused here.

aregng replied:
It is used in the error event.
```
  event::emit(
                ErrorInsufficientBalanceToRefund { refund_type, task_index, owner: task_owner, amount: refundable_amount }
            );
```

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -678,14 +813,255 @@ module supra_framework::automation_registry {
+        result
+    }
+
+    fun safe_unlock_locked_deposit(rb: &mut AutomationRefundBookkeeping, locked_deposit: u64, task_index: u64) {
+        if (rb.total_deposited_automation_fee > locked_deposit) {
+            rb.total_deposited_automation_fee = rb.total_deposited_automation_fee - locked_deposit;
+        } else {
+            event::emit(
+                ErrorUnlockTaskDepositFee { total_registered_deposit: rb.total_deposited_automation_fee, potential_locked_deposit: locked_deposit, task_index }
+            );
+        }
+
+    }
```
Do we maintain at all times that `balance` in the resource account is the same as `total_deposited_automation_fee`? If so, I believe `safe_refund` covers this and we do not need to check if `total_deposited_automation_fee > locked_deposit`.

aregng replied:
resource account will contain all the charged fees: `registration-fee`s , `automation-epoch-fee`s, and also `deposit fee`s.
From all of them the last 2 have possibility to be refunded. It is also possible to unlock the deposit without refunding, in case if user's balance is not enough to pay the automation-fee.
So I assume we need to check here extra,  not to cause overflow issue during epoch transition if there are any bugs.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -678,14 +813,255 @@ module supra_framework::automation_registry {
+        vector::for_each(ids, |task_index| {
+            let task = enumerable_map::get_value_ref(&automation_registry.tasks, task_index);
+
+            // Drop or activate task for this current epoch.
```
```suggestion
  
```

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -678,14 +813,255 @@ module supra_framework::automation_registry {
+
+    }
+
+    fun safe_fee_refund(resource_signer: &signer, resouce_address: address, task_index: u64, task_owner: address, refundable_fee: u64):  bool {
+        let result = safe_refund(resource_signer, resouce_address, task_index, task_owner, refundable_fee, EPOCH_FEE);
+        event::emit(
+            TaskFeeRefund { task_index, owner: task_owner, amount: refundable_fee }
+        );
+        result
+    }
```
I am wondering if just a single `safe_refund` will be sufficient or not. It seems `safe_deposit_refund` also does something same.

aregng replied:
If we do not want to distinguish between refunds then we can have a single event for all type of refunds, otherwise now we are emitting 2 different events on successful refund based on the refund-type.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -1079,6 +1601,7 @@ module supra_framework::automation_registry {
         automation_registry.gas_committed_for_next_epoch = committed_gas;
         let task_index = automation_registry.current_index;
 
+        let deposit_automtion_fee = automation_fee_cap_for_epoch;
```
```suggestion
        let deposit_automation_fee = automation_fee_cap_for_epoch;
```

Commit:
Addressed review comments

Commit:
Added temporary means to switch between 2 implementations of epoch handling

Commit:
Fixed after rebase

Commit:
Cleaned up automation_registry from obsolate code and fixed unit tests

Commit:
Did self-reivew and strengthend checks on safe refund logic

- Updated tests and added more
- Removed redundant code

Commit:
Fixed backward-compatibility issues of framework upgrade

- Optimized code

Commit:
Emitted event related to active tasks of the new epoch

Commit:
Fixed comments and types

nizam-supraoracles commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -1262,6 +1678,17 @@ module supra_framework::automation_registry {
                 };
+                let result = safe_unlock_locked_deposit(
+                    refund_bookkeeping,
+                    task.locked_fee_for_next_epoch,
```
Here we need to refund `deposit_refund` instead `task.locked_fee_for_next_epoch`

aregng replied:
This function simply unlocks the locked deposit, it does not refund anything,
Refund of the epoch-fee and deposited fee happens later when total_refund_fee (containing aggregated refunds, line 1693) is transferred from registry account to user account.

nizam-supraoracles commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -3254,15 +3809,237 @@ module supra_framework::automation_registry {
+        check_account_balance(get_registry_fee_address(), expected_registry_balance);
+        assert!(remaining_epoch_locked_fees == expected_total_locked, 4);
+
+        // Assume there is no enough balance to refund the epoch fee in regiatry account.
```
`regiatry` -> `registry`

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -98,6 +104,13 @@ module supra_framework::automation_registry {
+    const DEPOSIT_EPOCH_FEE: u8 = 0;
+    const EPOCH_FEE: u8 = 1;
+
+    /// Defines refund factor for refunds of deposit fees with penalty
```
```suggestion
    /// Defines divisor for refunds of deposit fees with penalty
    /// Factor of `2` suggests that `1/2` of the deposit will be refunded.
```

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -293,21 +376,31 @@ module supra_framework::automation_registry {
     struct DisabledRegistrationEvent has drop, store {}
 
     /// Represents the fee charged for an automation task execution and some additional information.
-    struct AutomationTaskFee has drop {
+    struct AutomationTaskFeeMeta has drop {
         task_index: u64,
         owner: address,
         fee: u64,
+        automation_fee_cap: u64,
+        expiry_time: u64,
+        max_gas_amount: u64,
+        locked_deposit_fee: u64
     }
```
It seems that an existing `struct` is being changed. This is not allowed in upgrade. @aregng

aregng replied:
`AutomationTaskFee` is still present and now it is used in scope of tests only, so I moved it to be closer to tests, line 1969

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -293,21 +376,31 @@ module supra_framework::automation_registry {
+    struct IntermediateStateOfEpochChange {
+        removed_tasks: vector<u64>,
+        gas_committed_for_new_epoch: u256,
+        gas_committed_for_next_epoch: u64,
```
Why is `gas_committed_for_new_epoch` is `u256` but `gas_committed_for_next_epoch` is `u64`? Wouldn't `u64` be sufficiently large for each of them? @aregng

aregng replied:
Here it can be updated to `u64` , but for some reason we missed that fact and current `AutomationRegistry` resource has `gas_committed_for_this_epoch` as  `u256` (as when congestion multiplier is calculated we upscale the values not to lose precision, but in resource it is downscaled version in `u256`),  and `gas_committed_for_next_epoch` has `u64`

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -556,11 +668,19 @@ module supra_framework::automation_registry {
     }
 
+    public fun initialize_refund_bookkeeping_resource(supra_framework: &signer) {
+        move_to(supra_framework, AutomationRefundBookkeeping {
```
Add a check that it is indeed `0x1` signer by `system_addresses::assert_supra_framework` so that others can not call this method. @aregng

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -571,178 +691,408 @@ module supra_framework::automation_registry {
-
-            automation_epoch_info.start_time = current_time;
-            automation_epoch_info.expected_epoch_duration = automation_epoch_info.epoch_interval;
+            finalize_epoch_change_for_feature_disabled_state(
+                automation_registry,
+                automation_epoch_info,
+                refund_bookkeeping,
+                current_time,
+                intermediate_state);
```
I am wondering what would happen for mainnet here. When we check `if (!features::supra_native_automation_enabled())`, we don't know if it was enabled earlier and now disabled or it was never enabled. If we assume that it was never enabled, would this cause an inconsistency? Since this code assumes existence of registry etc. @aregng

aregng replied:
The outcome/the resulting state is the same regardless whether feature flag was enabled or not.
The registry state will be cleaned up/reset. 
If the registry is not initialized, then it will not reach this point at all, as initialization state is checked at the beginning of the `on_new_epoch` function.
If the registry is already empty, then only epoch-info will be updated, the rest of the properties will remain empty/0.
 
Can you please provide more details what type of inconsistency we may encounter?

sjoshisupra replied:
@aregng , fair point. Well, I was wondering an attempt to access registry when it doesn't exist but since you mentioned we are unlikely to reach that point. At the moment, I can not think of other scenario for inconsistency. You can close this comment.

Commit:
Addressed review comments

aregng committed:
Updated threshold value for congestion to be exclusive

aregng committed:
[AN-Issue-1906,1907] Feature gated automation payload gas check upon registration (#231)

* [AN-Issue-1906,1907] Feature gated automation payload gas check upon registration

- Added dedicated VM status-codes for gas checks failures in scope of
  automation payload gas-parameters checks during automation task
  registration

* Fixed SUPAR_AUTOMATION_PAYLOAD_GAS_CHECK value in move context to be 90

---------

Co-authored-by: Aregnaz Harutyunyan <>
